lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1331,6 +1331,39 @@ void SPIRVToLLVM::transFunctionPointerCallArgumentAttributes(
   }
 }
 
+/// When spirv is generated from LLVM IR with opapque pointer enabled and then
+/// the spirv is translated to LLVM IR with typed pointer, function pointer is
+/// casted to i8* type in \p Ty and \p Initializer, which causes error in LLVM
+/// IR verifier. E.g.
+///   [1 x %0][%0 { i32 1, i8* bitcast (void ()* @ctor to i8*), i8* null }]
+/// This function removes the cast so that LLVM IR is valid.
+static void postProcGlobalCtorDtorTypeInit(Type *&Ty, Constant *&Initializer) {
+  auto *InitArr = cast<ConstantArray>(Initializer);
+  unsigned NumEltsInArr = InitArr->getType()->getNumElements();
+  assert(NumEltsInArr && "array is empty");
+  auto *CS = cast<ConstantStruct>(InitArr->getAggregateElement(0u));
+  assert(CS->getType()->getNumElements() == 3 && "expect 3 elements in struct");
```

Comment:
This new function is not added to validate, so I'm not sure you need new asserts here at all. It looks like the following logic can be just protected by if conditions, for example, to avoid adding implicit layer of validity checks into a new ad-hoc function that addresses ctor/dtor case. However, if other reviewers think that asserts are ok here and may remain as it is, I'd strongly suggest to consider better error messages, adding more context to the message to facilitate possible future testing/debugging efforts.